hazelcast/src/test/java/com/hazelcast/jet/impl/util/PMSTest.java

```diff
@@ -0,0 +1,424 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.jet.impl.util;
+
+import com.hazelcast.cluster.Address;
+import com.hazelcast.jet.Job;
+import com.hazelcast.jet.SimpleTestInClusterSupport;
+import com.hazelcast.jet.aggregate.AggregateOperations;
+import com.hazelcast.jet.config.JobConfig;
+import com.hazelcast.jet.core.AbstractProcessor;
+import com.hazelcast.jet.core.DAG;
+import com.hazelcast.jet.core.Edge;
+import com.hazelcast.jet.core.JobStatus;
+import com.hazelcast.jet.core.ProcessorMetaSupplier;
+import com.hazelcast.jet.core.ProcessorSupplier;
+import com.hazelcast.jet.core.Vertex;
+import com.hazelcast.jet.core.processor.Processors;
+import com.hazelcast.jet.datamodel.Tuple2;
+import com.hazelcast.test.HazelcastSerialClassRunner;
+import com.hazelcast.test.annotation.QuickTest;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.nio.file.Path;
+import java.security.Permission;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.Scanner;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import static com.hazelcast.jet.config.JobConfigArguments.KEY_REQUIRED_PARTITIONS;
+import static com.hazelcast.jet.core.Edge.between;
+import static com.hazelcast.jet.core.ProcessorMetaSupplier.forceTotalParallelismOne;
+import static java.util.Collections.nCopies;
+import static java.util.Collections.singleton;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+@RunWith(HazelcastSerialClassRunner.class)
+@Category(QuickTest.class)
+public class PMSTest extends SimpleTestInClusterSupport {
+    private static final String OUTPUT_FILE = "output.txt";
+    private File output;
+    private Scanner scanner;
+
+    @BeforeClass
+    public static void beforeClass() throws Exception {
+        initialize(3, null);
+    }
+
+    @Before
+    public void setUp() throws Exception {
+        output = new File(OUTPUT_FILE);
```

Comment:
Why a file? We usually use some static field / collection.

Reply:
It is multi-member test, and the most universal storage for multi-program output in my mind are files.

Reply:
Yes, but it still runs inside a single JVM, see e.g. `TestProcessors` - I would rather we follow that approach. The file approach will be flaky as there is no synchronization.

Reply:
https://github.com/hazelcast/hazelcast/pull/24689/commits/ef7632824c25ba3a2b5082f3182034bd9640a61a